oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+| Last Modified | 2016-12-12                                |
++---------------+-------------------------------------------+
+| Author        | Andy Armstrong <andya@edx.org>            |
++---------------+-------------------------------------------+
+| Arbiter       | Clinton Blackburn <clinton@edx.org>       |
++---------------+-------------------------------------------+
+| Status        | Draft                                     |
++---------------+-------------------------------------------+
+| Type          | Architecture                              |
++---------------+-------------------------------------------+
+| Created       | 2016-12-06                                |
++---------------+-------------------------------------------+
+
+Abstract
+========
+
+There is a long standing desire to allow developers to extend Open edX
+applications by providing plugins, rather than by modifying each application's
+code directly. XBlocks provide this ability for new types of courseware
+components, but courseware is only one of many aspects that a feature might need
+to provide (see `Feature Plugins for edX Platform`_ for an enumeration of such
+aspects for the edX platform).
+
+This proposal recommends that new user interfaces are plugged into applications
+by being defined as components that can be rendered into existing pages. These
+new components will be implemented using a new Django view subclass called
+`DjangoComponentView`. These views are rendered server-side, and an API is
+provided to allow clients to dynamically render a component if required.
+
+Motivation
+==========
+
+There are a number of driving use cases that act as motivation for supporting
+pluggable user interfaces.
+
+Use Case 1: LMS Course Tabs
+---------------------------
+
+A very common need for new features is to be able to add a new course tab for
+the learner, for example Discussions, Student Notes, and Teams. There is a
+supported pluggable entry point for course tabs (see `Adding a new course tab`_)
+but this requires that the feature add a new URL for the feature, and render the
+whole page at that URL using a Mako template. It would be preferable for a
+course tab to be able to register a pluggable user interface that will be
+presented as a tab, and for the implementor not to need to know how to implement
+the full page.
+
+There has been a long standing request to support rendering XBlocks as course
+tabs (`Allow xblocks to be shown as course tabs`_), but there are conceptual
+issues with this approach:
+
+* XBlocks must be configured and saved into the modulestore which doesn't make
+  much sense for course tabs that don't have their own state
+
+* Replacing a built-in tab such as "Course" would require a data migration for
+  every course if it was saved as an XBlock
+
+* It seems confusing to make an XBlock that should never be added into the
+  courseware, but which only exists to be shown as a tab
```

Comment:
The tabs should be part of the OLX. That was an artifact of how the platform was built. The component model didn't come in until some were already built.